### PR TITLE
Support other tree models #83

### LINKED ISSUE #83: support XGBoost and lightgbm?
It would be useful to support decision trees from gradient boosting machines. Should be a simple matter of creating a shadow tree from one of the trees in the boosted ensemble.  Interested in this one @tlapusan ?

### PR DESCRIPTION
**Pull request by tlapusan**
Hi @parrt,

This is the PR for xgboost and new architecture.
- it includes few changes about what we discussed yesterday, but is still work on progress on others.
- sklearn and xgboost notebooks are adapted to new changes, so you can take a look on them.

Then PR is ready, close #83

### CODE REVIEW

**Comment by parrt** [comment]
Looking good! I'll wait until you tell me it's ready. :)

**Comment by tlapusan** [comment]
Cool, I want to add few little changes, I will have time to add them tomorrow.

**Comment by tlapusan** [comment]
hi @parrt , we are ready to go. we can improve things afterwards and based on people feedback.

**Comment by parrt** [comment]
cool. i'll start looking.

**Comment by parrt** [comment]
Code looks great. i have a few review notes.

However, getting errors in the notebooks. Can you install then rerun them all? 

Also, the notebooks look like a junk drawer now. not sure we need the API in there. what about generating an API doc? Probably some python standard for doing this.

**Comment by tlapusan** [comment]
> Code looks great. i have a few review notes.
> 
> However, getting errors in the notebooks. Can you install then rerun them all?
       > _I've reran the notebooks (with your latest commits) and I've found one error in xgboost notebook. Which notebooks caused the errors ? (Forgot to adapt the examples and partitioning notebooks to the latest code, if you consider that we need to adapt them for this PR, I will do it)_
> Also, the notebooks look like a junk drawer now. not sure we need the API in there. what about generating an API doc? Probably some python standard for doing this.
       > _agree, I will remove them from notebook. Would be more helpful to have a documentation/API generated automatically by a library. Do you have any experience/preference with one of these  https://realpython.com/documenting-python-code/#documentation-tools-and-resources (btw, the article is very good for code documentation) ?_

File: testing/dtreeviz/models/conftest.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import pandas as pd
```
[the remark this answers is not in the document]
Comment (tlapusan) [reply testing/dtreeviz/models/conftest.py:1]: it is already in the testing module "testing/dtreeviz/models"... outside of main library module code "dtreeviz". did you mean something else about testing module ?

**Comment by parrt** [comment]
Will follow up later, but doxygen has always done well for me in other languages. https://www.doxygen.nl/manual/install.html

**Comment by parrt** [comment]
I uninstalled then ran setup.py for 1.0 and for dtreeviz_sklearn_visualisations.ipynb, i see:

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/opt/anaconda3/lib/python3.7/site-packages/matplotlib/projections/__init__.py in get_projection_class(projection)
     57     try:
---> 58         return projection_registry.get_projection_class(projection)
     59     except KeyError:

~/opt/anaconda3/lib/python3.7/site-packages/matplotlib/projections/__init__.py in get_projection_class(self, name)
     24         """
---> 25         return self._all_projection_types[name]
     26 

KeyError: '3d'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-69-17d285db2728> in <module>
----> 1 trees.rtreeviz_bivar_3D(dtr_bivar, dataset[features_reg_bivar], dataset[target_reg], features_reg_bivar, target_reg)

~/opt/anaconda3/lib/python3.7/site-packages/dtreeviz-1.0-py3.7.egg/dtreeviz/trees.py in rtreeviz_bivar_3D(tree_model, x_data, y_data, feature_names, target_name, class_names, tree_index, ax, fontsize, ticks_fontsize, fontname, azim, elev, dist, show, colors, markersize, n_colors_in_map)
    258     if ax is None:
    259         fig = plt.figure()
--> 260         ax = fig.add_subplot(111, projection='3d')
    261 
    262     colors = adjust_colors(colors)

~/opt/anaconda3/lib/python3.7/site-packages/matplotlib/figure.py in add_subplot(self, *args, **kwargs)
   1394         else:
   1395             projection_class, kwargs, key = \
-> 1396                 self._process_projection_requirements(*args, **kwargs)
   1397 
   1398             # try to find the axes with this key in the stack

~/opt/anaconda3/lib/python3.7/site-packages/matplotlib/figure.py in _process_projection_requirements(self, polar, projection, *args, **kwargs)
   1118 
   1119         if isinstance(projection, str) or projection is None:
-> 1120             projection_class = projections.get_projection_class(projection)
   1121         elif hasattr(projection, '_as_mpl_axes'):
   1122             projection_class, extra_kwargs = projection._as_mpl_axes()

~/opt/anaconda3/lib/python3.7/site-packages/matplotlib/projections/__init__.py in get_projection_class(projection)
     58         return projection_registry.get_projection_class(projection)
     59     except KeyError:
---> 60         raise ValueError("Unknown projection %r" % projection)
     61 
     62 

ValueError: Unknown projection '3d'
```

**Comment by tlapusan** [comment]
it's a little strange, the trees.rtreeviz_bivar_3D works for me.
bivar_3D

Do you have this error only for rtreeviz_bivar_3D ? Does it work from master branch ?

It seems to be a popular issue if I searched on google for "jupyter notebook matplotlib Unknown projection '3d'". Most probability a matplotlib version issue. 

These are my project dependencies versions :
pip freeze : 
appnope==0.1.0
attrs==19.3.0
backcall==0.1.0
bleach==3.1.3
colour==0.1.5
coverage==5.1
cycler==0.10.0
decorator==4.4.2
defusedxml==0.6.0
entrypoints==0.3
graphviz==0.13.2
importlib-metadata==1.5.0
ipykernel==5.2.0
ipython==7.13.0
ipython-genutils==0.2.0
ipywidgets==7.5.1
jedi==0.16.0
Jinja2==2.11.1
joblib==0.14.1
jsonschema==3.2.0
jupyter==1.0.0
jupyter-client==6.1.0
jupyter-console==6.1.0
jupyter-core==4.6.3
kiwisolver==1.1.0
MarkupSafe==1.1.1
**matplotlib==3.2.1**
mistune==0.8.4
more-itertools==8.2.0
nbconvert==5.6.1
nbformat==5.0.4
notebook==6.0.3
numpy==1.18.2
packaging==20.3
pandas==1.0.3
pandocfilters==1.4.2
parso==0.6.2
pexpect==4.8.0
pickleshare==0.7.5
pluggy==0.13.1
prometheus-client==0.7.1
prompt-toolkit==3.0.4
ptyprocess==0.6.0
py==1.8.1
Pygments==2.6.1
pyparsing==2.4.6
pyrsistent==0.15.7
pytest==5.4.1
pytest-cov==2.10.0
python-dateutil==2.8.1
pytz==2019.3
pyzmq==19.0.0
qtconsole==4.7.1
QtPy==1.9.0
scikit-learn==0.22.2.post1
scipy==1.4.1
Send2Trash==1.5.0
six==1.14.0
terminado==0.8.3
testpath==0.4.4
tornado==6.0.4
traitlets==4.3.3
wcwidth==0.1.9
webencodings==0.5.1
widgetsnbextension==3.5.1
xgboost==1.0.2
zipp==3.1.0

**Comment by parrt** [comment]
weird. maybe do an pip uninstall then setup.py install and restart the jupyter lab (i'm using lab not notebook btw). that way all is fresh and what I did.

**Comment by parrt** [comment]
I'm upgrading matplatlib... was 3.1.1

**Comment by parrt** [comment]
Ok, that notebook is working now. cool. colors doesn't though. i get:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-81abbfa92ee9> in <module>
----> 1 show_iris()

<ipython-input-3-4d5f8df9feeb> in show_iris(colors)
     15                     nbins=40, gtype='strip',
     16                     show={'splits','title'},
---> 17                     colors=colors)

~/opt/anaconda3/lib/python3.7/site-packages/dtreeviz-1.0-py3.7.egg/dtreeviz/trees.py in ctreeviz_univar(ax, x_train, y_train, feature_name, class_names, target_name, max_depth, min_samples_leaf, fontsize, fontname, nbins, gtype, show, colors)
    342 
    343     shadow_tree = ShadowDecTree(ct, x_train.reshape(-1, 1), y_train,
--> 344                                 feature_names=[feature_name], class_names=class_names)
    345 
    346     n_classes = shadow_tree.nclasses()

TypeError: Can't instantiate abstract class ShadowDecTree with abstract methods classes, criterion, get_children_left, get_children_right, get_class_weight, get_class_weights, get_feature_path_importance, get_features, get_max_depth, get_min_samples_leaf, get_node_criterion, get_node_feature, get_node_samples, get_node_split, get_prediction_value, get_score, get_thresholds, is_classifier, is_fit, nclasses, nnodes
```

**Comment by parrt** [comment]
Near bottom of xgboost notebook i get:

```
trees.rtreeviz_univar(xgb_shadow_reg_univar, dataset[features_reg_univar], dataset[target_reg_univar], features_reg_univar, target_reg_univar)
```

```
---------------------------------------------------------------------------
VisualisationNotYetSupportedError         Traceback (most recent call last)
<ipython-input-42-e56c79818b73> in <module>
----> 1 trees.rtreeviz_univar(xgb_shadow_reg_univar, dataset[features_reg_univar], dataset[target_reg_univar], features_reg_univar, target_reg_univar)

~/opt/anaconda3/lib/python3.7/site-packages/dtreeviz-1.0-py3.7.egg/dtreeviz/trees.py in rtreeviz_univar(tree_model, x_data, y_data, feature_names, target_name, class_names, tree_index, ax, fontsize, show, split_linewidth, mean_linewidth, markersize, colors)
    152 
    153     if 'title' in show:
--> 154         title = f"Regression tree depth {shadow_tree.get_max_depth()}, samples per leaf {shadow_tree.get_min_samples_leaf()},\nTraining $R^2$={shadow_tree.get_score()}"
    155         ax.set_title(title, fontsize=fontsize, color=colors['title'])
    156 

~/opt/anaconda3/lib/python3.7/site-packages/dtreeviz-1.0-py3.7.egg/dtreeviz/models/xgb_decision_tree.py in get_min_samples_leaf(self)
    196 
    197     def get_min_samples_leaf(self):
--> 198         raise VisualisationNotYetSupportedError("get_min_samples_leaf()", "XGBoost")

VisualisationNotYetSupportedError: get_min_samples_leaf() is not implemented yet for XGBoost
```

**Comment by parrt** [comment]
partitioning, for

```
t = rtreeviz_univar(ax,
                    X['WGT'], y,
                    max_depth=3,
                    feature_name='Vehicle Weight',
                    target_name='MPG',
                    fontsize=10,
                    colors={'scatter_edge': 'black'})
```

i get 

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-7a70b2bb45a2> in <module>
      6                     target_name='MPG',
      7                     fontsize=10,
----> 8                     colors={'scatter_edge': 'black'})
      9 plt.tight_layout()
     10 plt.show()

TypeError: rtreeviz_univar() got an unexpected keyword argument 'max_depth'
```

**Comment by parrt** [comment]
tree structure.

```
dtreeviz(dtc, dataset[features], dataset[target], features, target, class_names, instance_orientation="TD")
```

gives

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-18-595382d82fdc> in <module>
----> 1 dtreeviz(dtc, dataset[features], dataset[target], features, target, class_names, instance_orientation="TD")

~/opt/anaconda3/lib/python3.7/site-packages/dtreeviz-1.0-py3.7.egg/dtreeviz/trees.py in dtreeviz(tree_model, x_data, y_data, feature_names, target_name, class_names, tree_index, precision, orientation, instance_orientation, show_root_edge_labels, show_node_labels, show_just_path, fancy, histtype, highlight_path, X, max_X_features_LR, max_X_features_TD, label_fontsize, ticks_fontsize, fontname, colors, scale)
    780 
    781     shadow_tree = ShadowDecTree.get_shadow_tree(tree_model, x_data, y_data, feature_names, target_name, class_names,
--> 782                                                 tree_index)
    783     colors = adjust_colors(colors)
    784 

~/opt/anaconda3/lib/python3.7/site-packages/dtreeviz-1.0-py3.7.egg/dtreeviz/models/shadow_decision_tree.py in get_shadow_tree(tree_model, x_data, y_data, feature_names, target_name, class_names, tree_index)
    413             return tree_model
    414         elif isinstance(tree_model, (sklearn.tree.DecisionTreeRegressor, sklearn.tree.DecisionTreeClassifier)):
--> 415             return dtreeviz.models.sklearn_decision_trees.ShadowSKDTree(tree_model, x_data, y_data, feature_names,
    416                                                                         target_name, class_names)
    417         elif isinstance(tree_model, xgboost.core.Booster):

AttributeError: module 'dtreeviz.models' has no attribute 'sklearn_decision_trees'
```

**Comment by tlapusan** [comment]
> Near bottom of xgboost notebook i get:
> 
> ```
> trees.rtreeviz_univar(xgb_shadow_reg_univar, dataset[features_reg_univar], dataset[target_reg_univar], features_reg_univar, target_reg_univar)
> ```
> 
> ```
> ---------------------------------------------------------------------------
> VisualisationNotYetSupportedError         Traceback (most recent call last)
> <ipython-input-42-e56c79818b73> in <module>
> ----> 1 trees.rtreeviz_univar(xgb_shadow_reg_univar, dataset[features_reg_univar], dataset[target_reg_univar], features_reg_univar, target_reg_univar)
> 
> ~/opt/anaconda3/lib/python3.7/site-packages/dtreeviz-1.0-py3.7.egg/dtreeviz/trees.py in rtreeviz_univar(tree_model, x_data, y_data, feature_names, target_name, class_names, tree_index, ax, fontsize, show, split_linewidth, mean_linewidth, markersize, colors)
>     152 
>     153     if 'title' in show:
> --> 154         title = f"Regression tree depth {shadow_tree.get_max_depth()}, samples per leaf {shadow_tree.get_min_samples_leaf()},\nTraining $R^2$={shadow_tree.get_score()}"
>     155         ax.set_title(title, fontsize=fontsize, color=colors['title'])
>     156 
> 
> ~/opt/anaconda3/lib/python3.7/site-packages/dtreeviz-1.0-py3.7.egg/dtreeviz/models/xgb_decision_tree.py in get_min_samples_leaf(self)
>     196 
>     197     def get_min_samples_leaf(self):
> --> 198         raise VisualisationNotYetSupportedError("get_min_samples_leaf()", "XGBoost")
> 
> VisualisationNotYetSupportedError: get_min_samples_leaf() is not implemented yet for XGBoost
> ```

I do have this error also. This is because I didn't find a way to get min_sample_leaf from xgboost. Not sure if xgboost has this hyperparameter like decision tree from sklearn. work in progress on this.

**Comment by tlapusan** [comment]
For the rest of the errors, I assume they are from partitioning, examples and tree_structure_examples notebooks. They are not yet adapter to the new architecture. 
I've asked you in a previous comment it they are mandatory for this PR. If they are mandatory, we can adapt them, but I assume that the answer will be "they are" :)

**Comment by tlapusan** [comment]
examples notebook should work now

**Comment by tlapusan** [comment]
we can remove tree_structure_examples notebook because we can find its content also in dtreeviz_sklearn_visualisations notebook

**Comment by parrt** [comment]
howdy! Ah. sorry. i'm stuck in other stuff so you're only getting 60% of my brain when I look.  Thanks for cleaning stuff up.  If `min_sample_leaf` is a problem, let's just make a note there and keep that cell as raw not code then further examples proceed.

for examples.ipynb...i thought we decided to fold that stuff into the other notebooks?

if partitioning classification visualization isn't ready we should wait so library is still cohesive and functional since people likely use it. how much work is to to update? Maybe I can try it.

**Comment by tlapusan** [comment]
hi @parrt , you are right, we decided to fold the examples.ipynb into the other notebooks(sklearn, xgboost)...but it's a little confusing for me right now : 
  - in which notebook to put the examples.ipynb visualisations ? sklearn or xgboost ? examples.ipynb contains mainly dtreeviz() visualizations, only on different datasets.... sklearn and xgboost also contains dtreeviz(). 
  - I'm not sure where should we include these visualizations, top, middle of the sklearn, xgboost notebooks ? 
  - should we keep the datasets from examples.ipynb ? (like boston, iris, digits, etc datasets)

I assume i can adapt the partitioning classification viz tomorrow. Just let me know if you want to implement these yourself.

**Comment by parrt** [comment]
I assume you know the code better so if you can adapt to the partitioning that would be great. I wonder if we shouldn't start again and organize notebooks then cut-and-paste to make it much better.

**Comment by tlapusan** [comment]
> I assume you know the code better so if you can adapt to the partitioning that would be great. I wonder if we shouldn't start again and organize notebooks then cut-and-paste to make it much better.

agree with better organizing the notebooks

**Comment by parrt** [comment]
maybe we leave the notebook reorg for after the release? we just have to make sure the examples that exist work :)  can delete those that don't

**Comment by tlapusan** [comment]
hi @parrt, please check the partitioning examples. They should work now.

**Comment by parrt** [comment]
In colors notebook, i'm getting after uninstall / reinstall:

```
<ipython-input-4-81abbfa92ee9> in <module>
----> 1 show_iris()

<ipython-input-3-4d5f8df9feeb> in show_iris(colors)
     15                     nbins=40, gtype='strip',
     16                     show={'splits','title'},
---> 17                     colors=colors)

TypeError: ctreeviz_univar() got an unexpected keyword argument 'max_depth'
```

and

```
<ipython-input-15-6032d1711aea> in <module>
     14 
     15 print('unchanged')
---> 16 change_heatmap(None)
     17 
     18 print('all changes')

<ipython-input-15-6032d1711aea> in change_heatmap(colors)
      7                                               feature_names=['Vehicle Weight', 'Horse Power'],
      8                                               fontsize=14,
----> 9                                               colors=colors)
     10     plt.show()
     11 

TypeError: rtreeviz_bivar_heatmap() got an unexpected keyword argument 'max_depth'
```

**Comment by parrt** [comment]
I removed some empty cells from a few notebooks and pushed it back.

**Comment by tlapusan** [comment]
ohh, I did't observe the colors notebook.... :( I see that it contains a lot of visualizations. I can take a look over them in this weekend.

**Comment by tlapusan** [comment]
hi @parrt, pls check the colors notebook. let's hope everything is working 😄

**Comment by parrt** [comment]
Ok, it works!  Merging!  Then we can test some more. I will merge but then I have an appointment for the next couple of hours, so let's play around with it before releasing.

**Comment by parrt** [comment]
Great job, @tlapusan !!!

**Comment by tlapusan** [comment]
@parrt, totally agree, we should play with it before releasing, we need to also update the readme.md file for the latest changes.

**Pull request merged**